sync2/migration.go

```diff
@@ -0,0 +1,275 @@
+	for i, device := range devices {
+		device.AccessToken, err = decrypt(device.AccessTokenEncrypted, key)
+		if err != nil {
+			return
+		}
+		logger.Info().Msgf("%4d/%4d migrating device %s", i+1, len(devices), device.AccessTokenHash)
+		err = migrateDevice(txn, whoamiClient, &device)
+		if err != nil {
+			return
+		}
+	}
+
+	return
+}
+
+func migrateDevice(txn *sqlx.Tx, whoamiClient Client, device *oldDevice) (err error) {
+	gotUserID, gotDeviceID, err := whoamiClient.WhoAmI(device.AccessToken)
+	if err != nil {
+		return err
+	}
+	// Sanity check the user ID from the HS matches our records
+	if gotUserID != device.UserID {
+		return fmt.Errorf(
+			"/whoami response was for the wrong user. Queried for %s, but got response for %s",
+			device.UserID, gotUserID,
+		)
+	}
+
+	// For these first four tables:
+	// - use the actual device ID instead of the access token hash, and
+	// - ensure a user ID is set.
+	err = exec(
+		txn,
+		`UPDATE syncv3_sync2_devices SET user_id = $1, device_id = $2 WHERE device_id = $3`,
+		expectOneRowAffected,
+		gotUserID, gotDeviceID, device.AccessTokenHash,
+	)
+	if err != nil {
+		return
+	}
+
+	err = exec(
+		txn,
+		`UPDATE syncv3_to_device_messages SET user_id = $1, device_id = $2 WHERE device_id = $3`,
+		expectAnyNumberOfRowsAffected,
+		gotUserID, gotDeviceID, device.AccessTokenHash,
+	)
+	if err != nil {
+		return
+	}
+
+	err = exec(
+		txn,
+		`UPDATE syncv3_to_device_ack_pos SET user_id = $1, device_id = $2 WHERE device_id = $3`,
+		expectAtMostOneRowAffected,
+		gotUserID, gotDeviceID, device.AccessTokenHash,
+	)
+	if err != nil {
+		return
+	}
+
+	err = exec(
+		txn,
+		`UPDATE syncv3_device_data SET user_id = $1, device_id = $2 WHERE device_id = $3`,
+		expectAtMostOneRowAffected,
+		gotUserID, gotDeviceID, device.AccessTokenHash,
+	)
+	if err != nil {
+		return
+	}
+
+	// Confusingly, the txns table used to store access token hashes under the user_id
+	// column. Write the actual user ID to the user_id column, and the actual device ID
+	// to the device_id column.
+	err = exec(
+		txn,
+		`UPDATE syncv3_txns SET user_id = $1, device_id = $2 WHERE user_id = $3`,
+		expectAtMostOneRowAffected,
+		gotUserID, gotDeviceID, device.AccessTokenHash,
+	)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+func exec(txn *sqlx.Tx, query string, checkRowsAffected func(ra int64) bool, args ...any) error {
+	res, err := txn.Exec(query, args...)
+	if err != nil {
+		return err
+	}
+	ra, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if !checkRowsAffected(ra) {
+		return fmt.Errorf("Failed checkRowsAffected: got %d", ra)
+	}
+	return nil
+}
+
+func expectOneRowAffected(ra int64) bool          { return ra == 1 }
+func expectAnyNumberOfRowsAffected(ra int64) bool { return true }
+func expectAtMostOneRowAffected(ra int64) bool    { return ra == 0 || ra == 1 }
+
+func finish(txn *sqlx.Tx) (err error) {
+	_, err = txn.Exec(`
+		ALTER TABLE syncv3_sync2_devices
+		DROP COLUMN v2_token_encrypted,
```

Comment:
Why drop this?

Reply:
I thought the point was to move the tokens to the tokens table. Is there any reason to keep it around in the devices table? I suppose that could reflect that access token that a particular poller is using right now in the moment... but I'm not sure how we'd use that information.

Reply:
I just didn't see a tokens table when I looked.

Reply:
Oh bugger, I forgot to insert the new rows! Arg.